grpc.go

```diff
@@ -81,9 +81,53 @@ func handleError(wait bool, err error) error {
 	return err
 }
 
-// Online resources, nbResources specifies the maximum number of resources to online.
-// If nbResources is <= 0 then there is no limit and all resources are connected.
+// Offline resources, nbResources specifies the maximum number of resources to online.
+// If nbResources is <= 0 then there is no limit and all resources are disconnected.
 // Returns the number of resources connected.
+func offlineResources(resource onlineResource, nbResources int32) (uint32, error) {
+	// Get the bitmap of the online cpus
+	onlineCpuBitmapPath := filepath.Join(resource.sysfsOnlinePath, "online")
+	status, err := ioutil.ReadFile(onlineCpuBitmapPath)
+	if err != nil {
+		return 0, err
+	}
+
+	// Get from the last online cpu
+	strArr := strings.Split(strings.Replace(string(status), ",", "-", -1), "-")
+	strLastOnlineCpu := strings.Trim((strArr[len(strArr)-1]), "\n\t ")
+	lastOnlineCpu, err := strconv.Atoi(strLastOnlineCpu)
+	if err != nil {
+		return 0, err
+	}
+	if lastOnlineCpu == 0 {
+		agentLog.Warnf("Can not offline cpu 0")
```

Comment:
- I think this would be clearer as:
  ```go
  agentLog.Warnf("offlining last available CPU disallowed")
  ```

- Paranoia: handle `lastOnlineCpu < 0` case?

  This is partially handled by the `for` loop below, but if `lastOnlineCpu == -1`, the function will return `(-1, nil)` rather than an error.

Reply:
ok

Reply:
> * I think this would be clearer as:
>   ```go
>   agentLog.Warnf("offlining last available CPU disallowed")
>   ```
> * Paranoia: handle `lastOnlineCpu < 0` case?
>   This is partially handled by the `for` loop below, but if `lastOnlineCpu == -1`, the function will return `(-1, nil)` rather than an error.

func offlineResources(resource onlineResource, nbResources int32) (uint32, error)
The first return value should be uint32.  And in offlineCPUResources
`        var count uint32
        for i := uint32(0); i < cpuStateChangeMaxTries; i++ {
                r, err := offlineResources(resource, int32(nbCpus-count))
                if err != nil {
                        return err
                }
                count += r`
So if return (-1,nil), count might be decreased exceptionally?